crit/decode.go

```diff
@@ -0,0 +1,270 @@
+			return err
+		}
+		// Create proto struct to hold payload
+		payload, err := images.ProtoHandler(img.Magic)
+		if err != nil {
+			return err
+		}
+		payloadSize := uint64(binary.LittleEndian.Uint32(sizeBuf))
+		payloadBuf := make([]byte, payloadSize)
+		if _, err := f.Read(payloadBuf); err != nil {
+			return err
+		}
+		if err := proto.Unmarshal(payloadBuf, payload); err != nil {
+			return err
+		}
+		entry := CriuEntry{Message: payload}
+		if decodeExtra != nil {
+			extraPayload, err := decodeExtra(f, payload, noPayload)
+			if err != nil {
+				return err
+			}
+			entry.Extra = extraPayload
+		}
+		img.Entries = append(img.Entries, &entry)
+	}
+	return nil
+}
+
+func (img *CriuImage) decodePagemap(f *os.File) error {
+	sizeBuf := make([]byte, 4)
+	// First entry is pagemap head
+	var payload proto.Message = &images.PagemapHead{}
+	// Read payload size and payload until EOF
+	for {
+		n, err := f.Read(sizeBuf)
+		if n == 0 {
+			if err == io.EOF {
+				break
+			}
+			return err
+		}
+
+		payloadSize := uint64(binary.LittleEndian.Uint32(sizeBuf))
+		payloadBuf := make([]byte, payloadSize)
+		if _, err := f.Read(payloadBuf); err != nil {
+			return err
+		}
+		if err := proto.Unmarshal(payloadBuf, payload); err != nil {
+			return err
+		}
+		entry := CriuEntry{Message: payload}
+		img.Entries = append(img.Entries, &entry)
+		// Create struct for next entry
+		payload = &images.PagemapEntry{}
+	}
+	return nil
+}
+
+func (img *CriuImage) decodeGhostFile(f *os.File, noPayload bool) error {
+	sizeBuf := make([]byte, 4)
+	if _, err := f.Read(sizeBuf); err != nil {
+		return err
+	}
+	// Create proto struct for primary entry
+	payload := &images.GhostFileEntry{}
+	payloadSize := uint64(binary.LittleEndian.Uint32(sizeBuf))
+	payloadBuf := make([]byte, payloadSize)
+	if _, err := f.Read(payloadBuf); err != nil {
+		return err
+	}
+	if err := proto.Unmarshal(payloadBuf, payload); err != nil {
+		return err
+	}
+	entry := CriuEntry{Message: payload}
+
+	if payload.GetChunks() {
+		img.Entries = append(img.Entries, &entry)
+		for {
+			n, err := f.Read(sizeBuf)
+			if n == 0 {
+				if err == io.EOF {
+					break
+				}
+				return err
+			}
+			// Create proto struct for chunk
+			payload := &images.GhostChunkEntry{}
+			payloadSize := uint64(binary.LittleEndian.Uint32(sizeBuf))
+			payloadBuf := make([]byte, payloadSize)
+			if _, err := f.Read(payloadBuf); err != nil {
+				return err
+			}
+			if err := proto.Unmarshal(payloadBuf, payload); err != nil {
+				return err
+			}
+			entry = CriuEntry{Message: payload}
+			if noPayload {
+				if _, err = f.Seek(int64(payload.GetLen()), 1); err != nil {
+					return err
+				}
+			} else {
+				extraBuf := make([]byte, payload.GetLen())
+				if _, err := f.Read(extraBuf); err != nil {
+					return err
+				}
+				entry.Extra = base64.StdEncoding.EncodeToString(extraBuf)
+			}
+			img.Entries = append(img.Entries, &entry)
+		}
+	} else {
+		if noPayload {
+			if _, err := f.Seek(0, 2); err != nil {
```

Comment:
It would be good to add a comment here. Why do we call `f.Seek(0, 2)`?

Reply:
The `(0, 2)` indicates "seek to an offset of 0 relative to the end of the file". https://pkg.go.dev/os#File.Seek
I've added a comment explaining this :+1: